ring/partition_instance_ring.go

```diff
@@ -0,0 +1,123 @@
+package ring
+
+import (
+	"fmt"
+	"time"
+)
+
+type PartitionRingReader interface {
+	PartitionRing() *PartitionRing
+}
+
+// PartitionInstanceRing holds a partitions ring and a instances ring, and provide functions
+// to look up the intersection of the two (e.g. healthy instances by partition).
+type PartitionInstanceRing struct {
+	// partitionsRingReader and partitionsRing are mutually exclusive.
+	partitionsRingReader PartitionRingReader
+	partitionsRing       *PartitionRing
+
+	instancesRing    *Ring
+	heartbeatTimeout time.Duration
+}
+
+func NewPartitionInstanceRing(partitionsRingWatcher PartitionRingReader, instancesRing *Ring, heartbeatTimeout time.Duration) *PartitionInstanceRing {
+	return &PartitionInstanceRing{
+		partitionsRingReader: partitionsRingWatcher,
+		instancesRing:        instancesRing,
+		heartbeatTimeout:     heartbeatTimeout,
+	}
+}
+
+func newStaticPartitionInstanceRing(partitionsRing *PartitionRing, instancesRing *Ring, heartbeatTimeout time.Duration) *PartitionInstanceRing {
+	return &PartitionInstanceRing{
+		partitionsRing:   partitionsRing,
+		instancesRing:    instancesRing,
+		heartbeatTimeout: heartbeatTimeout,
+	}
+}
+
+func (r *PartitionInstanceRing) PartitionRing() *PartitionRing {
+	if r.partitionsRingReader != nil {
+		return r.partitionsRingReader.PartitionRing()
+	}
+
+	return r.partitionsRing
+}
+
+func (r *PartitionInstanceRing) InstanceRing() *Ring {
+	return r.instancesRing
+}
+
+// GetReplicationSetsForOperation returns one ReplicationSet for each partition in the ring.
+// A ReplicationSet is returned for every partition in ring. If there are no healthy owners
+// for a partition, an error is returned.
+func (r *PartitionInstanceRing) GetReplicationSetsForOperation(op Operation) ([]ReplicationSet, error) {
+	partitionsRing := r.PartitionRing()
+	partitionsRingDesc := partitionsRing.desc
+
+	if len(partitionsRingDesc.Partitions) == 0 {
+		return nil, ErrEmptyRing
+	}
+
+	now := time.Now()
+	result := make([]ReplicationSet, 0, len(partitionsRingDesc.Partitions))
+
+	for partitionID := range partitionsRingDesc.Partitions {
+		ownerIDs := partitionsRing.UnsafePartitionOwnerIDs(partitionID)
```

Comment:
Instead of `Unsafe` prefix, perhaps should we call the *other* method to indicate that it returns `Copy` (ie. we would have `PartitionOwnerIDs` and `PartitionOwnerIDsCopy`)? I get what you're trying to say with it, but it seems weird that we're using "unsafe" methods.

Reply:
Addressed in https://github.com/grafana/dskit/pull/479/commits/e33bb87dc4c8eb13e91aeff9d22a8f35a4246950